ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -866,112 +916,156 @@ validateCandidate lgrDB tracer cfg varInvalid blockCache
+  -> m (Maybe (Either (CandidateSuffix blk) (CandidateSuffixAndLedger blk)))
+futureCheck tracer varInvalid varFutureBlocks CheckInFuture { checkInFuture }
+            candidateSuffixAndLedger =
+    checkInFuture validatedFragment >>= \case
+
+      (candChain', []) ->
+        -- If no headers are in the future, then the fragment must be untouched
+        assert (AF.headPoint candChain == AF.headPoint candChain') $
+        return $ Just $ Right candidateSuffixAndLedger
+
+      (candChain', inFuture) -> do
+        let (exceedClockskew, inNearFuture) =
+              partition InFuture.inFutureExceedsClockskew inFuture
+        -- Record future blocks
+        unless (null inNearFuture) $ do
+          let futureHeaders = InFuture.inFutureHeader <$> inNearFuture
+              futureBlocks  = Map.fromList
+                [ (headerHash hdr, hdr) | hdr <- futureHeaders ]
+          atomically $ modifyTVar varFutureBlocks $ flip Map.union futureBlocks
+          trace $ CandidateContainsFutureBlocks candChain futureHeaders
+
+        -- Record any blocks exceeding the clockskew as invalid
+        unless (null exceedClockskew) $ do
+          let invalidHeaders = InFuture.inFutureHeader <$> exceedClockskew
+              invalidBlocks  = Map.fromList
+                [ (headerHash hdr, info)
+                | hdr <- invalidHeaders
+                , let reason = InFutureExceedsClockSkew (headerRealPoint hdr)
+                      info   = InvalidBlockInfo reason (blockSlot hdr)
+                ]
+          atomically $ modifyTVar varInvalid $ \(WithFingerprint invalid fp) ->
+            WithFingerprint (Map.union invalid invalidBlocks) (succ fp)
+          trace $
+            CandidateContainsFutureBlocksExceedingClockSkew
+              -- Trace the original @candChain@, as it contains the headers
+              -- from the future
+              candChain
+              invalidHeaders
+
+        -- Use the trimmed 'AnchoredFragment' to create an 'CandidateSuffix'
+        -- with the same rollback parameter
+        let mbCandSuffix = CS.mkRollback
+              (CS.getRollback candidateSuffix)
+              candChain'
+        return $ Left <$> mbCandSuffix
+  where
+    trace = traceWith tracer
+
+    candidateSuffix :: CandidateSuffix blk
+    candidateSuffix = CS.getCandidateSuffix candidateSuffixAndLedger
+
+    candChain :: AnchoredFragment (Header blk)
+    candChain = CS.getSuffix candidateSuffix
+
+    validatedFragment :: ValidatedFragment blk (LedgerState blk)
+    validatedFragment =
+      ledgerState . LgrDB.ledgerDbCurrent <$>
+      CS.toValidatedFragment candidateSuffixAndLedger
+
+-- | Validate a candidate chain using 'ledgerValidate' and 'futureCheck'.
+--
+-- Note: this function doesn't take a 'ChainDbEnv' because it is also used on
+-- startup, before we have set up the 'ChainDbEnv.
+validateCandidate
+  :: ( IOLike m
+     , LedgerSupportsProtocol blk
+     , HasCallStack
+     )
+  => LgrDB m blk
```

Comment:
Introduce `ChainSelEnv`. Can the `ChainDbEnv` contain the `ChainSelEnv`.

Reply:
> Introduce `ChainSelEnv`.

Done

> Can the `ChainDbEnv` contain the `ChainSelEnv`.

I did not do this, it was more convenient to store the `BlockCache` and the current `ChainAndLedger` in the `ChainSelEnv`, which makes it impossible to store the `ChainSelEnv` in the `ChainDbEnv`.